// client/src/context/types.js
export const REGISTER_SUCCESS = 'REGISTER_SUCCESS';
export const REGISTER_FAIL = 'REGISTER_FAIL';
export const USER_LOADED = 'USER_LOADED ';
export const AUTH_ERROR = 'AUTH_ERROR';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAIL = 'LOGIN_FAIL';
export const LOGOUT = 'LOGOUT';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';
export const SET_ALERT = 'SET_ALERT';
export const REMOVE_ALERT = 'REMOVE_ALERT';
export const UPDATE_DETAILS = 'UPDATE_DETAILS';
export const UPDATE_DETAILS_FAIL = 'UPDATE_DETAILS_FAIL';
export const CREATE_PRODUCT = 'CREATE_PRODUCT';
export const GET_PRODUCTS = 'GET_PRODUCTS';
export const PRODUCTS_FAIL = 'PRODUCTS_FAIL';
export const GET_PRODUCT = 'GET_PRODUCT';
export const REMOVE_PRODUCT = 'REMOVE_PRODUCT';
export const ADD_TO_BASKET = 'ADD_TOBASKET';
export const GET_BASKET = 'GET_BASKET';
export const DELETE_BASKET_ITEM = 'DELETE_BASKET_ITEM';
export const FILTER_BY_CATEGORY = 'FILTER_BY_CATEGORY';
export const CLEAR_CURRENT_FILTER = 'CLEAR_CURRENT_FILTER';
export const FILTER_BY_RATING = 'FILTER_BY_RATING';
export const RATE_PRODUCT = 'RATE_PRODUCT';
export const SEARCH_PRODUCTS = 'SEARCH_PRODUCTS';
export const CLEAR_SEARCH_RESULTS = 'CLEAR_SEARCH_RESULTS';
export const CREATE_CHECKOUT_SESSION = 'CREATE_CHECKOUT_SESSION';
